cns/nodesubnet/ip_fetcher.go

```diff
@@ -10,46 +10,130 @@ import (
+	ticker             TickProvider
+	tickerInterval     time.Duration
+	consumer           IPConsumer
+	minRefreshInterval time.Duration
+	maxRefreshInterval time.Duration
 }
 
-func NewIPFetcher(nmaClient InterfaceRetriever, queryInterval time.Duration) *IPFetcher {
+// NewIPFetcher creates a new IPFetcher. If minInterval is 0, it will default to 4 seconds.
+// If maxInterval is 0, it will default to 1024 seconds (or minInterval, if it is higher).
+func NewIPFetcher(
+	client InterfaceRetriever,
+	consumer IPConsumer,
+	minInterval time.Duration,
+	maxInterval time.Duration,
+) *IPFetcher {
+	if minInterval == 0 {
+		minInterval = DefaultMinRefreshInterval
+	}
+
+	if maxInterval == 0 {
+		maxInterval = DefaultMaxRefreshInterval
+	}
+
+	maxInterval = max(maxInterval, minInterval)
+
 	return &IPFetcher{
-		ipFectcherClient:         nmaClient,
-		secondaryIPQueryInterval: queryInterval,
+		intfFetcherClient:  client,
+		consumer:           consumer,
+		minRefreshInterval: minInterval,
+		maxRefreshInterval: maxInterval,
+		tickerInterval:     minInterval,
+	}
+}
+
+func (c *IPFetcher) UpdateFetchIntervalForNoObservedDiff() {
+	c.tickerInterval = min(c.tickerInterval*2, c.maxRefreshInterval) //nolint:gomnd // doubling interval
+
+	if c.ticker != nil {
+		c.ticker.Reset(c.tickerInterval)
 	}
 }
 
-func (c *IPFetcher) RefreshSecondaryIPsIfNeeded(ctx context.Context) (ips []netip.Addr, err error) {
-	// If secondaryIPQueryInterval has elapsed since the last fetch, fetch secondary IPs
-	if time.Since(c.secondaryIPLastRefreshTime) < c.secondaryIPQueryInterval {
-		return nil, ErrRefreshSkipped
+func (c *IPFetcher) UpdateFetchIntervalForObservedDiff() {
+	c.tickerInterval = c.minRefreshInterval
+
+	if c.ticker != nil {
+		c.ticker.Reset(c.tickerInterval)
+	}
+}
+
+func (c *IPFetcher) Start(ctx context.Context) {
+	go func() {
+		// Do an initial fetch
+		err := c.RefreshSecondaryIPs(ctx)
+		if err != nil {
+			log.Printf("Error refreshing secondary IPs: %v", err)
+		}
+
+		if c.ticker == nil {
+			c.ticker = NewTimedTickProvider(c.tickerInterval)
+		}
+
+		defer c.ticker.Stop()
+
+		for {
+			select {
+			case <-c.ticker.C():
+				err := c.RefreshSecondaryIPs(ctx)
+				if err != nil {
+					log.Printf("Error refreshing secondary IPs: %v", err)
+				}
+			case <-ctx.Done():
+				log.Println("IPFetcher stopped")
+				return
+			}
+		}
+	}()
+}
+
+// Fetch IPs from NMAgent and pass to the consumer
+func (c *IPFetcher) RefreshSecondaryIPs(ctx context.Context) error {
+	response, err := c.intfFetcherClient.GetInterfaceIPInfo(ctx)
+	if err != nil {
+		return errors.Wrap(err, "getting interface IPs")
+	}
+
+	if len(response.Entries) == 0 {
+		return errors.New("no interfaces found in response from NMAgent")
 	}
 
-	c.secondaryIPLastRefreshTime = time.Now()
-	response, err := c.ipFectcherClient.GetInterfaceIPInfo(ctx)
+	primaryIP, secondaryIPs := flattenIPListFromResponse(&response)
+	err = c.consumer.UpdateIPsForNodeSubnet(primaryIP, secondaryIPs)
```

Comment:
what primary IP is used for?

Reply:
You might recall that our plan is to use an artificial Nodesubnet NC to hold the IPs. NCs expect a primary IP to be present. Now, we could leave it empty in Nodesubnet, but it seemed better to populate it since we are fetching it anyway.

Reply:
ok got it. Lets add this as comment

Reply:
Ended up removing it, because it's ultimately unnecessary and was making endpoint state restoration complicated.